app/src/ui/discard-changes/index.tsx

```diff
@@ -78,9 +78,9 @@ export class DiscardChanges extends React.Component<
 
         <DialogFooter>
           <ButtonGroup destructive={true}>
-            <Button type="submit">Cancel</Button>
+            <Button type="submit">Discard Changes</Button>
             <Button onClick={this.discard}>
```

Comment:
This event handler can be moved up to `Dialog.onSubmit` due to the button switch.

Reply:
So this is gonna be `Dialog.onSubmit = {this.discard}`?

Reply:
Plus, I'm gonna remove ` <Button onClick={this.discard}>` from here right? @shiftkey

Reply:
So I stumbled upon something in here I wasn't expecting - `Dialog` requires a `onDismissed` handler to be set.

The behaviour we had before:

 - the cancel button acts as submit, which by default dismisses the dialog due to firing this `onDismissed` handler
 - the discard button handles it's own click event, and then calls `this.props.onDismissed()` itself

So some back story in here before we go further:

 - note the `destructive={true}` on the parent `ButtonGroup` element
 - we brought this in because on macOS we wanted to flip these buttons around because of a report in https://github.com/desktop/desktop/issues/1311 that we should flip the buttons around due to the macOS style guide
 - the code to apply this is done here https://github.com/desktop/desktop/pull/1315
 - I think this change will affect that in some weird way

I get the feeling that the proper fix for this should be done in `ButtonGroup`, rather than reordering the DOM like this.

Let me wrap my head around why these dialogs are different on Windows to our others and report back.

Reply:
Let's see if we can do this without changing the DOM around. What I think we need to do:

 - for a ButtonGroup marked `destructive={true}`
   - when not on macOS, flip the order of the buttons
   - invert the styles (submit should be styled like a button, regular button should be styled like a submit)

That second part does feel weird, but it might be easier than fighting with how our dialogs are setup. Also, we don't add any CSS classes here `destructive={true}` for the dialog buttons so we can style them (cc @desktop/design here to tell me if they have any better ideas).

Reply:
@shiftkey Thank you so much for the detailed explanation, as I'm reading through I have the same feelings about the changes that you suggested. Let's see what @desktop/design has to say.